File: app/src/main/java/com/estebanlamas/planesradar/data/AircraftDataRepository.java
package com.estebanlamas.planesradar.data;

import com.estebanlamas.planesradar.data.mapper.AircraftMapper;
import com.estebanlamas.planesradar.data.remote.AdsbExchangeApi;
import com.estebanlamas.planesradar.data.remote.response.AircraftListResponse;
import com.estebanlamas.planesradar.domain.model.AircraftsDetected;
import com.estebanlamas.planesradar.domain.repository.AircraftRepository;

import javax.inject.Inject;

import rx.Observable;
import rx.functions.Func1;

/**
 * Created by esteban on 26/12/17
 */

public class AircraftDataRepository extends BaseRepository implements AircraftRepository{

    @Inject
    public AircraftDataRepository(AdsbExchangeApi adsbExchangeApi) {
        super(adsbExchangeApi);
    }

    public Observable<AircraftsDetected> getAircraftList() {
        return adsbExchangeApi.getAircraftList().flatMap(new Func1<AircraftListResponse, Observable<AircraftsDetected>>() {
            @Override
            public Observable<AircraftsDetected> call(AircraftListResponse aircraftListResponse) {
                return Observable.just(AircraftMapper.mapResponse(aircraftListResponse));
            }
        });
    }
}

File: app/src/main/java/com/estebanlamas/planesradar/data/AircraftDataRepository.kt
package com.estebanlamas.planesradar.data

import com.estebanlamas.planesradar.data.mapper.AircraftMapper
import com.estebanlamas.planesradar.data.remote.AdsbExchangeApi
import com.estebanlamas.planesradar.domain.model.AircraftsDetected
import com.estebanlamas.planesradar.domain.repository.AircraftRepository
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import javax.inject.Inject

/**
 * Created by esteban on 26/12/17
 */

class AircraftDataRepository
@Inject constructor(
    adsbExchangeApi: AdsbExchangeApi
) : BaseRepository(adsbExchangeApi), AircraftRepository {

    companion object {
        const val REFRESH_MILLISECONDS = 3000L
    }

    override fun aircraftList(): Flow<AircraftsDetected> = flow {
        while(true) {
            emit(AircraftMapper.mapResponse(adsbExchangeApi.aircraftList()))
            delay(REFRESH_MILLISECONDS)
        }
    }
}
